### PR TITLE
Add Radio Scanner MechComp component

### PR DESCRIPTION
**Pull request by glassofmilk**
[ENHANCEMENT]

## About the PR 
This adds a MechComp component that can monitor voice radio stations and output a MechComp packet with the speaker and the message. Secure channels, such as Command, Security, Engineering, etc, cannot be monitored - only the channels that intercoms can be tuned to. Unless the speaker is using a voice changer, it will recognize their voice.

## Why's this needed? 
Finally, you can create a doomsday device that will only trigger if someone says "antiquing" over the radio!

## Changelog

```
(u)glassofmilk:
(*)Added a Radio Scanner Component to the MechComp Dispenser. It can monitor voice radio channels and output a packet accordingly.
```

### CODE REVIEW

**Comment by ZeWaka** [comment]
Why not just have it use the spoken name heard over radio?

**Comment by glassofmilk** [comment]
Largely because I had a hard time understanding /obj/item/device/radio.dm. If you could point me to what you mean, I'd be happy to make the change.

File: code/obj/item/device/radio.dm [hunk 1]
```diff
@@ -330,6 +330,9 @@ Green Wire: <a href='?src=\ref[src];wires=[WIRE_TRANSMIT]'>[src.wires & WIRE_TRA
 
 	var/list/heard_flock = list() // heard by flockdrones/flockmind
 
+	for (var/obj/item/mechanics/radioscanner/I in connection.devices) //MechComp radio scanner
```
Comment (pali6) [inline code/obj/item/device/radio.dm:333]: I'd probably merge the two loops into one to be more efficient.

File: code/WorkInProgress/MechanicMadness.dm [hunk 2]
```diff
@@ -2072,6 +2072,69 @@ var/list/mechanics_telepads = new/list()
...
+		frequency = new_frequency
+		radio_connection = radio_controller.add_object(src, "[frequency]")
+
+	proc/hear_radio(mob/M as mob, msg, lang_id)
+		if (level == 2) return
+		var/message = msg[2]
+		if (lang_id in list("english", ""))
+			message = msg[1]
+		message = strip_html(html_decode(message))
+		var/heardname = M.real_name
```
Comment (pali6) [inline code/WorkInProgress/MechanicMadness.dm:2126]: Use `get_heard_name()` instead.

File: code/WorkInProgress/MechanicMadness.dm [hunk 3]
```diff
@@ -2072,6 +2072,69 @@ var/list/mechanics_telepads = new/list()
...
+
+	proc/setfreq(var/datum/mechanicsMessage/input)
+		var/newfreq = text2num(input.signal)
+		if (!newfreq) return
+		set_frequency(newfreq)
+		return
+
+	proc/set_frequency(new_frequency)
+		if (!radio_controller) return
+		new_frequency = max(R_FREQ_MINIMUM, min(new_frequency, R_FREQ_MAXIMUM))
```
Comment (pali6) [inline code/WorkInProgress/MechanicMadness.dm:2115]: I think for regular radios you can't set even values manually if I recall correctly? This should probably behave similarly.

Comment (glassofmilk) [reply code/WorkInProgress/MechanicMadness.dm:2115]: Actually, you can! Regular headsets and intercoms can all be set to any frequency between R_FREQ_MINIMUM and R_FREQ_MAXIMUM. This range includes intercom channels, but no department specific frequencies. The default general chat, R_FREQ_DEFAULT (1459), falls within this range.

Comment (pali6) [reply code/WorkInProgress/MechanicMadness.dm:2115]: When I tried I was unable to set an intercom / headset to an even frequency.

File: code/obj/item/device/radio.dm (hunk 1 above)
Comment (glassofmilk) [reply code/obj/item/device/radio.dm]: Ok, this is updated

File: code/WorkInProgress/MechanicMadness.dm (hunk 2 above)
Comment (pali6) [reply code/WorkInProgress/MechanicMadness.dm:2126]: Actually nevermind, that's a different thing. Our code for radio stuff sucks.

File: code/WorkInProgress/MechanicMadness.dm (hunk 3 above)
Comment (ZeWaka) [reply code/WorkInProgress/MechanicMadness.dm:2115]: > When I tried I was unable to set an intercom / headset to an even frequency.

IIRC, that's intentional.

Comment (glassofmilk) [reply code/WorkInProgress/MechanicMadness.dm]: I'm dumb, I missed the word "even" in the original comment. Looks like using sanitize_frequency will handle that as well as keeping the new frequency within the allowed range. This issue should be resolved

**Pull request merged**